### PR TITLE
Snippet feedback

### PR DESCRIPTION
**Pull request by Jbarget**
- Creates global dialog `dialog` in model
- holds `snippetFeedback` in model
- adjusts feedback request to send `snippetFeedback`
- server validation adjusted for `suggestedCategories`
- `suggestedCategories` sent to SearchAPI
- renames endpoint from `reject` to `suggest` (and in Elm land too)

screen shot 2018-07-20 at 17 07 07
screen shot 2018-07-20 at 17 06 57

### CODE REVIEW

File: client/src/Update/Update.elm [hunk 1]
```diff
@@ -468,7 +541,7 @@ update msg model =
                 , Navigation.modifyUrl ("/#/query?" ++ newQueryString)
                 )
 
-        SnippetRejectClick ( snippetId, resultIndex ) ->
+        SnippetSuggestClick ( snippetId, resultIndex ) ->
```
Comment (mantagen) [inline client/src/Update/Update.elm]: using resultIndex?

File: client/src/Dialog/SnippetFeedback.elm [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+                  )
+                , ( "disabled o-30", isDisabled )
+                ]
+    in
+        section [ class "flex justify-center items-center flex-column mt3" ]
+            [ h1 [ class "f5 w-90 pt1 pb2 mh1 bb b--black-20 mb1" ] [ text "Re-Train RegSimple" ]
+            , button
+                [ type_ "button"
+                , class baseButtonClass
+                , onClick (SnippetSuggestClick snippetData)
```
Comment (mantagen) [inline client/src/Dialog/SnippetFeedback.elm]: Here we wanna be sending `focusElementId`  too? for when you submit, your focus isn't suddenly at the end of the document ..;

Comment (mantagen) [reply client/src/Dialog/SnippetFeedback.elm]: (if that's the course of action we choose)

File: client/src/Dialog/SnippetFeedback.elm [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+            , form [ class "w-100" ]
+                [ div [ class "flex justify-around items-center" ]
+                    [ activitySelect activityMenuFeedbackModel
+                    , categorySelect categoryMenuFeedbackModel
+                    ]
+                ]
+            , button
+                [ disabled isDisabled
+                , submitButtonClass
+                , onClick (SnippetSuggestClick snippetData)
```
Comment (mantagen) [inline client/src/Dialog/SnippetFeedback.elm]: likewise

File: client/src/Model/Model.elm [hunk 4]
```diff
@@ -94,6 +103,18 @@ type alias Model =
...
     , loginCodeResponse : WebData User
     , session : Session
+    , dialog : DialogType
+    , snippetFeedback : SnippetFeedback
+    }
+
+
+initialSnippetFeedback : SnippetFeedback
+initialSnippetFeedback =
+    { activityId = ""
```
Comment (mantagen) [inline client/src/Model/Model.elm]: Objection to this being `Nothing`?

File: client/src/Data/DataTypes.elm [hunk 5]
```diff
@@ -23,6 +23,54 @@ type InputAlignment
...
+type alias Category =
+    AppDataItem
+
+
+type alias SnippetFeedback =
+    { activityId : ActivityId
+    , activityMenuOpen : Bool
+    , categoryIds : List CategoryId
+    , categoryMenuOpen : Bool
+    , snippetData : ( SnippetId, Int )
```
Comment (mantagen) [inline client/src/Data/DataTypes.elm]: What is this `Int` here?

File: server/__test__/router/feedback-snippet.test.js [hunk 6]
```diff
@@ -83,4 +83,25 @@ describe('GET /feedback/snippet/:snippetId/:feedbackType', () => {
...
+      .expect(200)
+      .end(done);
+  });
+  test('400 if invalid suggestedCategories passed', done => {
+    return request(app)
+      .put(
+        '/api/feedback/snippet/id123/suggest?countries[]=GB&categories[]=aml-authority'
+      )
+      .set('Cookie', [adminCookie])
+      .send({ suggestedCategories: '' })
```
[the remark this answers is not in the document]
Comment (Jbarget) [reply server/__test__/router/feedback-snippet.test.js]: what would we want to happen in this case? At the moment it comes back with a 200

**Pull request merged**